packages/es-dev-server/src/config.ts

```diff
@@ -0,0 +1,282 @@
+  babelModernExclude?: string[];
+  // files excluded from module transfomration
+  babelModuleExclude?: string[];
+  // babel config to use, this is useful when you want to provide a
+  // babel config from a tool, and don't want to require all users to use the same babel config
+  babelConfig?: TransformOptions;
+  // callback called before the server is
+  // started, the returned promise is awaited
+  onServerStart?: (config: InternalConfig) => void | Promise<void>;
+}
+
+// Internal config data structure, constructed from the public config
+export interface InternalConfig {
+  // Server configuration
+  port?: number;
+  hostname?: string;
+  openBrowser: boolean;
+  openPath: string;
+  appIndex?: string;
+  appIndexDir?: string;
+  basePath?: string;
+  rootDir: string;
+  logStartup: boolean;
+  customMiddlewares: Middleware[];
+  responseTransformers: ResponseTransformer[];
+  onServerStart?: (config: InternalConfig) => void | Promise<void>;
+
+  // Development help
+  watch: boolean;
+  logErrorsToBrowser: boolean;
+  watchDebounce: number;
+  http2: boolean;
+  sslKey?: string;
+  sslCert?: string;
+
+  // Code transformation
+  nodeResolve: boolean | NodeResolveOptions;
+  polyfillsLoaderConfig?: Partial<PolyfillsLoaderConfig>;
+  readUserBabelConfig: boolean;
+  compatibilityMode: string;
+  compress: boolean | CompressOptions;
+  customBabelConfig?: TransformOptions;
+  fileExtensions: string[];
+  babelExclude: string[];
+  babelModernExclude: string[];
+  babelModuleExclude: string[];
+}
+
+/**
+ * Creates dev server config with default settings
+ */
+export function createConfig(config: Partial<Config>): InternalConfig {
+  const {
+    babel = false,
+    babelConfig,
+    babelExclude = [],
+    babelModernExclude = [],
+    babelModuleExclude = [],
+    basePath,
+    compress = true,
+    fileExtensions: fileExtensionsArg,
+    hostname,
+    http2 = false,
+    logStartup,
+    open = false,
+    port,
+    sslCert,
+    sslKey,
+    watch = false,
+    logErrorsToBrowser = false,
+    polyfillsLoader,
+    responseTransformers = [],
+    debug = false,
+    nodeResolve: nodeResolveArg = false,
+    dedupeModules,
+    moduleDirs = ['node_modules', 'web_modules'],
+    preserveSymlinks = false,
+    onServerStart,
+  } = config;
+
+  if (debug) {
+    setDebug(true);
+  }
+
+  let { compatibility = compatibilityModes.AUTO } = config;
+
+  if (compatibility === 'modern' || compatibility === 'all') {
+    /* eslint-disable-next-line no-console */
+    console.warn(
+      '[es-dev-server] Compatibility mode "modern" and "all" are deprecated, and combined into "auto".' +
+        '"auto" mode is turned on by default.',
+    );
+    compatibility = compatibilityModes.AUTO;
+  }
+
+  if (!Object.values(compatibilityModes).includes(compatibility)) {
+    throw new Error(`Unknown compatibility mode: ${compatibility}`);
+  }
+
+  // middlewares used to be called customMiddlewares
+  // @ts-ignore
```

Comment:
y tho

Reply:
It's a piece of backwards compatibility code, I don't want it to show up in the interface. I'll cast to any though, that's probably better.